### ISSUE TITLE
CSV reader crashes on 1.1TB file with 64GB memory available

### ISSUE THREAD

**Issue by k0rsakov** [duckdb/duckdb#8507]
### What happens?

I have `.csv.gz` a file that weighs 1.1 tb.
It has 18,724,386,915 lines

I want to work with him using duckdb.

But I keep getting the same error.

duckdb kills the process by 100%, the database size is growing, but the process is being killed and I can't use the data.

### To Reproduce

1. Use a file with 18,724,386,915 lines.
2. Download duckdb 0.8.1 to the virtual machine (duckdb_cli-linux-amd64.zip ).
3. Through the CLI, call the command `./duck click.dcdb` and get into interactive mode.
4. Execute the command:
```sql
create table click as 
select * from read_csv_auto('../data/file.csv.gz ') 
where activity_kind = 'click';
```
5. Wait for a while. It takes me about 17 hours. And the process will self-destruct.

An interesting point: the progress reaches 100% within 7-8 hours, is at 100% and the size of `click.duckdb` is gradually increasing. After about 10 hours, duckdb writes "Killed".

8. I run the command `./duck click.duckdb.`
9. I am writing a 
```sql
select count(*) from click;
```
10. I get an exception: 
```log
Error: Catalog Error: Table with name click does not exist!
Did you mean "information_schema.columns"?
LINE 1: select count(*) from click;
```

### OS:

Ubuntu 22.04.2 LTS (GNU/Linux 5.15.0-78-generic x86_64)

### DuckDB Version:

0.8.1

### DuckDB Client:

CLI

### Full Name:

Ivan Korsakov

### Affiliation:

By yourself

### Have you tried this on the latest `master` branch?

I have tested with a release build (and could not test with a master build)

### Have you tried the steps to reproduce? Do they include all relevant data and configuration? Does the issue you report still appear there?

- [X] Yes, I have

**Comment by szarnyasg** [comment]
Hi @k0rsakov, thanks for reporting this. A few quick questions to clarify the problem:
1) How much memory does the machine have?
2) How many threads are available on the machine?
3) What is the [decompressed size](https://superuser.com/questions/619591/how-can-i-get-the-uncompressed-size-of-gzip-file-without-actually-decompressing) of the CSV file?
4) How complex is the schema? Did you try defining the schema before loading the data?

**Comment by l1t1** [comment]
3. `gzip -l file.csv.gz`

**Comment by k0rsakov** [comment]
@l1t1  

```bash
compressed        uncompressed  ratio uncompressed_name
1106060442713          3564653820 -30928.6% file.csv
```

**Comment by l1t1** [comment]
what about use `7z l  file.csv.gz`

**Comment by szarnyasg** [comment]
Try passing `parallel=false` to the CSV reader, i.e.:
```
select * from read_csv_auto('../data/file.csv.gz', parallel=false)
```
Of course, this will be slower but it will help limit the memory consumption.
cc @pdet

**Comment by k0rsakov** [comment]
> Try passing `parallel=false` to the CSV reader, i.e.:
> 
> ```
> select * from read_csv_auto('../data/file.csv.gz', parallel=false)
> ```
> 
> Of course, this will be slower but it will help limit the memory consumption. cc @pdet

@szarnyasg Fine. Thank you. I will try this option and report the result later.

**Comment by k0rsakov** [comment]
@szarnyasg I ran the script with the `parallel=false` setting, the query really started to work slower.

But he was killed faster, literally an hour later. The download progress was 4%, previously reached 100% and worked longer.

**Comment by l1t1** [comment]
decompression and insert in one time maybe cost more. 
try gunzip first.

**Comment by k0rsakov** [comment]
@l1t1 I'm afraid I won't have enough disk space...

I have only 1.5tb, and the file already weighs 1.1tb.

**Comment by szarnyasg** [comment]
Thanks, I'll grab a large(-ish) CSV file and try to reproduce this issue. It will take some time of course.

**Comment by k0rsakov** [comment]
@szarnyasg Thanks for the feedback, I'll be waiting for news from you

**Comment by szarnyasg** [comment]
Do you have an indication of how selective the `where activity_kind = 'click'` filter is, i.e. what percentage of the tuples satisfy this filter?

**Comment by k0rsakov** [comment]
@szarnyasg 

- total 18,724,386,915
- no clicks 231,316,586
- clicks 18,493,070,329

Total clicks 98.76%

**Comment by szarnyasg** [comment]
@k0rsakov I'm afraid you will then run out of disk space: 1 GB of `.csv.gz` typically requires approx. 1.0-1.2 GB of space in DuckDB's binary storage format.

That said, there may also be an out of memory issue, but the disk space is definitely going to be a problem.

Depending on your workload, you can try running your queriers directly on the `.csv.gz` file. This can work e.g. for aggregation-heavy workloads.

**Comment by k0rsakov** [comment]
@szarnyasg 

I have two disks on my virtual machine. One contains a 1.1 tb file.

And on another disk, I'm just creating a database from this file. 

Therefore, I have no limit on creating a database. Where I create a database there is 800 GB.

In the past, an attempt at 100% of the database took up about 350 kb.

**Comment by l1t1** [comment]
you may use  `sed` of linux to get click.csv

**Comment by szarnyasg** [comment]
@k0rsakov
> Where I create a database there is 800 GB.
That may not be enough to store the database for a 1.1TB `.csv.gz` file if the selectivity is this high (98%+).

> In the past, an attempt at 100% of the database took up about 350 kb.
That seems to be wrong: 350 kb can barely store the metadata of this data set.

**If** there is sufficient disk space, then the loading should work. Is there a chance you can share your schema / data set with me?

@l1t1 
> you may use sed of linux to get click.csv
`sed` uses a lot of temporary disk space. Grep could perform this operation but only on decompressed data.

**Comment by k0rsakov** [comment]
Hi.

I also get problems with solving my question.

I expanded the space on the VM and was able to unzip mine .csv.gz -> .csv.

Its size was approximately 5.5tb

I used the same constructs as before to create the database. But also got a failure.

The creation of the table was "killed" several times for 30% success

At the time of the "killing" of the process, the database weighed approximately 445Gb.

At the moment I have refused to use duckdb to solve my question and am looking for other options to solve my question

I just wanted to share my successes and failures in this matter, maybe it will help you improve the process of creating a database and it will help someone.

thanks
